src/box/lua/call.c

```diff
@@ -92,61 +94,239 @@ get_call_serializer(void)
 	}
 }
 
-/**
- * A helper to find a Lua function by name and put it
- * on top of the stack.
- */
+/* Lexer for procedure resolution. */
+struct lexer {
+	/* Source string. */
+	const char *src;
+	/* Length of string. */
+	size_t src_len;
+	/* Current lexer offset in bytes. */
+	size_t offset;
+	/* Current lexer offset in symbols. */
+	size_t symbol_count;
+};
```

Comment:
An external state seems like an overkill. Can you do the whole parsing in `box_lua_find` without polluting the file namespace?

Also, we don't need to handle unicode symbols here - at least we didn't handle them before this patch.

Reply:
> Also, we don't need to handle unicode symbols here - at least we didn't handle them before this patch.

What makes you think so? Please study [this](https://github.com/tarantool/tarantool/blob/1d6043fac53bebd539ee5882f6073767eea753bd/test/box/function1.test.lua#L101-L129) test and the related GitHub tickets.

Reply:
> An external state seems like an overkill. Can you do the whole parsing in box_lua_find without polluting the file namespace?

It doesn't seem like an overkill, inlining everything looks incomprehensible.

Reply:
Up to @alyapunov.

Reply:
I propose alternative solution: https://github.com/tarantool/tarantool/pull/8660
I refactored and commented `box_lua_find` and simply add bracket support. The function is just few lines longer than the original one, but I hope it became more simple and understandable. It also naturally support negative and fractional numbers in brackets. I also stole your test cases but changed the test that it runs much faster.

Reply:
#8660 was merged.